File: ruby/calculator_cli_app/src/validations/args_validation.rb
# frozen_string_literal: true
# rbs_inline: enabled

module Validations
  # Argument-shape checks for CalculatorCliApp::Application.
  module ArgsValidation
    # @rbs args_size: Integer
    # @rbs seed: String
    # @rbs num: Integer?
    # @rbs return: void
    def validate!(args_size, seed, num)
      if args_size > 2
        puts 'Too many arguments'
        exit 1
      elsif args_size.zero?
        puts 'No argument provided'
        exit 1
      elsif args_size == 1
        puts 'Provide both seed and n'
        exit 1
      elsif !seed.instance_of?(String) || num.nil?
        puts 'Provide seed as string and n as number'
        exit 1
      end
    end
  end
end

File: python/calculator_cli_app/src/validations/args_validation.py
import sys


class ArgsValidation:
    """Argument-shape checks for Application; intended to be mixed in."""

    def validate(self, args_size: int, seed: str | None,
                 n: int | None) -> None:
        message = self._validation_error(args_size, seed, n)
        if message is None:
            return
        print(message)
        sys.exit(1)

    # private

    def _validation_error(self, args_size: int, seed: str | None,
                          n: int | None) -> str | None:
        if args_size > 2:
            return 'Too many arguments'
        if args_size == 0:
            return 'No argument provided'
        if args_size == 1:
            return 'Provide both seed and n'
        if not isinstance(seed, str) or n is None:
            return 'Provide seed as string and n as number'
        return None
